server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/identity/PostgresCustomIdentityModule.java

```diff
@@ -0,0 +1,69 @@
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.postgres.identity;
+
+import static org.apache.james.jmap.postgres.identity.PostgresCustomIdentityModule.PostgresCustomIdentityTable.TABLE;
+
+import java.util.UUID;
+
+import org.apache.james.backends.postgres.PostgresModule;
+import org.apache.james.backends.postgres.PostgresTable;
+import org.jooq.Field;
+import org.jooq.Record;
+import org.jooq.Table;
+import org.jooq.impl.DSL;
+import org.jooq.impl.SQLDataType;
+
+public interface PostgresCustomIdentityModule {
+    interface PostgresCustomIdentityTable {
+        Table<Record> TABLE_NAME = DSL.table("custom_identity");
+
+        Field<String> USERNAME = DSL.field("username", SQLDataType.VARCHAR(255).notNull());
+        Field<UUID> ID = DSL.field("id", SQLDataType.UUID.notNull());
+        Field<String> NAME = DSL.field("name", SQLDataType.VARCHAR(255).notNull());
+        Field<String> EMAIL = DSL.field("email", SQLDataType.VARCHAR(255).notNull());
+        Field<String[]> REPLY_TO = DSL.field("reply_to", SQLDataType.VARCHAR.getArrayDataType().notNull());
+        Field<String[]> BCC = DSL.field("bcc", SQLDataType.VARCHAR.getArrayDataType().notNull());
+        Field<String> TEXT_SIGNATURE = DSL.field("text_signature", SQLDataType.VARCHAR(255).notNull());
+        Field<String> HTML_SIGNATURE = DSL.field("html_signature", SQLDataType.VARCHAR(255).notNull());
+        Field<Integer> SORT_ORDER = DSL.field("sort_order", SQLDataType.INTEGER.notNull());
+        Field<Boolean> MAY_DELETE = DSL.field("may_delete", SQLDataType.BOOLEAN.notNull());
+
+        PostgresTable TABLE = PostgresTable.name(TABLE_NAME.getName())
+            .createTableStep(((dsl, tableName) -> dsl.createTableIfNotExists(tableName)
+                .column(USERNAME)
+                .column(ID)
+                .column(NAME)
+                .column(EMAIL)
+                .column(REPLY_TO)
+                .column(BCC)
+                .column(TEXT_SIGNATURE)
+                .column(HTML_SIGNATURE)
+                .column(SORT_ORDER)
+                .column(MAY_DELETE)
+                .constraint(DSL.primaryKey(USERNAME, ID))))
+            .supportsRowLevelSecurity()
+            .build();
+    }
+
+    PostgresModule MODULE = PostgresModule.builder()
+        .addTable(TABLE)
+        .build();
```

Comment:
on more Index for only `username` column?

Reply:
I think it is covered in index (USERNAME, ID)

Reply:
Link for such a claim?

Reply:
"A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns"

https://www.postgresql.org/docs/current/indexes-multicolumn.html

Reply:
I had a quick lab for testing:
Here is the result of query `explain select * from custom_identity where  username='user9811';`:

- Query where `username` with only `custom_identity_pkey`
![Screenshot_20240130_185832](https://github.com/apache/james-project/assets/81145350/871e3127-d3e0-4e01-b91f-e6cddce2b166)

- Query where `username` with `custom_identity__index_username`  (Which I tried to create with only `username` column)
![image](https://github.com/apache/james-project/assets/81145350/dff93c23-097d-42a1-abc3-55cbf4a15bdb)

We can see the `cost` of `custom_identity__index_username` is better

_________

// Script for generating sample data if anyone interested:

```sql
INSERT INTO custom_identity (username, id, name, email, reply_to, bcc, text_signature, html_signature, sort_order, may_delete, domain)
SELECT
    'user' || generate_series(1, 10000) AS username,
    uuid_generate_v4(),
    'Name' || generate_series(1, 10000) AS name,
    'email' || generate_series(1, 10000) || '@example.com' AS email,
    '{}' AS reply_to,
    '{}' AS bcc,
    'Text Signature' AS text_signature,
    'HTML Signature' AS html_signature,
    null AS sort_order,
    true AS may_delete,
    'example.com' AS domain
FROM generate_series(1, 10000);
```

Reply:
Probably in this case, we have a lot of ids per username so it is better to create index for usename only. Index added